### PR TITLE
2nd try at using a vanity GCR name

### PR DESCRIPTION
**Pull request by thockin**
The 2nd commit here is the changes relative to the reverted PR.  Please focus review attention on that.

This is the 2nd attempt.  The previous try (#57573) was reverted while we
figured out the regional mirrors (oops).
    
New plan: k8s.gcr.io is a read-only facade that auto-detects your source
region (us, eu, or asia for now) and pulls from the closest.  To publish
an image, push k8s-staging.gcr.io and it will be synced to the regionals
automatically (similar to today).  For now the staging is an alias to
gcr.io/google_containers (the legacy URL).
    
When we move off of google-owned projects (working on it), then we just
do a one-time sync, and change the google-internal config, and nobody
outside should notice.
    
We can, in parallel, change the auto-sync into a manual sync - send a PR
to "promote" something from staging, and a bot activates it.  Nice and
visible, easy to keep track of.

xref https://github.com/kubernetes/release/issues/281

TL;DR:
  *  The new `staging-k8s.gcr.io` is where we push images.  It is literally an alias to `gcr.io/google_containers` (the existing repo) and is hosted in the US.
  * The contents of `staging-k8s.gcr.io` are automatically synced to `{asia,eu,us)-k8s.gcr.io`.
  * The new `k8s.gcr.io` will be a read-only alias to whichever regional repo is closest to you.
  * In the future, images will be promoted from `staging` to regional "prod" more explicitly and auditably.

 ```release-note
Use "k8s.gcr.io" for pulling container images rather than "gcr.io/google_containers".  Images are already synced, so this should not impact anyone materially.
    
Documentation and tools should all convert to the new name. Users should take note of this in case they see this new name in the system.
```

### CODE REVIEW

File: CHANGELOG-1.4.md [hunk 1]
```diff
@@ -177,15 +177,15 @@ filename | sha256 hash
 ### Other notable changes
 
 * kube-apiserver now drops unneeded path information if an older version of Windows kubectl sends it. ([#44586](https://github.com/kubernetes/kubernetes/pull/44586), [@mml](https://github.com/mml))
-* Bump gcr.io/google_containers/glbc from 0.8.0 to 0.9.2. Release notes: [0.9.0](https://github.com/kubernetes/ingress/releases/tag/0.9.0), [0.9.1](https://github.com/kubernetes/ingress/releases/tag/0.9.1), [0.9.2](https://github.com/kubernetes/ingress/releases/tag/0.9.2) ([#43098](https://github.com/kubernetes/kubernetes/pull/43098), [@timstclair](https://github.com/timstclair))
+* Bump k8s.gcr.io/glbc from 0.8.0 to 0.9.2. Release notes: [0.9.0](https://github.com/kubernetes/ingress/releases/tag/0.9.0), [0.9.1](https://github.com/kubernetes/ingress/releases/tag/0.9.1), [0.9.2](https://github.com/kubernetes/ingress/releases/tag/0.9.2) ([#43098](https://github.com/kubernetes/kubernetes/pull/43098), [@timstclair](https://github.com/timstclair))
```
Comment (cblecker) [inline CHANGELOG-1.4.md]: Is it worth updating historial changelogs?

**Comment by cblecker** [comment]
FYI cross is failing across the board: #57822

File: CHANGELOG-1.4.md (hunk 1 above)
Comment (thockin) [reply CHANGELOG-1.4.md]: I waffle on this.  The image may eventually cease to exist at the old URL?  Or maybe we simply say it won't?

Comment (cblecker) [reply CHANGELOG-1.4.md]: My thought is, the application configuration and manifests for, say 1.4, will still be looking for the image at `gcr.io/google_containers/glbc`. Modifying the changelog, doesn't go back and create a new patch release of 1.4 updating the config/manifests to look at the new place.

I'd say it would be worth doing for any version we are actually going to go back and change. If this is just for 1.10 forward, then I'd suggest leaving the historical change logs be.

**Comment by dekkagaijin** [comment]
Pushes should now be made to `staging-k8s.gcr.io`. `k8s.gcr.io` will be turned into a read-only alias for the local replicas (i.e. `(asia|eu|us)-k8s.gcr.io`).

File: CHANGELOG-1.4.md (hunk 1 above)
Comment (thockin) [reply CHANGELOG-1.4.md]: ACK

File: cluster/common.sh [hunk 2]
```diff
@@ -478,7 +478,7 @@ function stage-images() {
 
   local docker_cmd=("docker")
 
-  if [[ "${KUBE_DOCKER_REGISTRY}" == "gcr.io/"* ]]; then
+  if [[ "${KUBE_DOCKER_REGISTRY}" =~ "gcr.io/" ]]; then
     local docker_push_cmd=("gcloud" "docker")
```
[the remark this answers is not in the document]
Comment (thockin) [reply cluster/common.sh]: Once staging becomes a default, we only need the -- part, right?

**Comment by thockin** [comment]
Un-holding this.  I think it works now.  Let's see what tests say.  PTAL

File: api/openapi-spec/swagger.json [hunk 3]
```diff
@@ -75018,7 +75018,7 @@
     ],
     "properties": {
      "names": {
-      "description": "Names by which this image is known. e.g. [\"gcr.io/google_containers/hyperkube:v1.0.7\", \"dockerhub.io/google_containers/hyperkube:v1.0.7\"]",
+      "description": "Names by which this image is known. e.g. [\"k8s.gcr.io/hyperkube:v1.0.7\", \"dockerhub.io/google_containers/hyperkube:v1.0.7\"]",
```
Comment (cblecker) [inline api/openapi-spec/swagger.json]: Does the dockerhub.io/google_containers make sense here? (even if it's still just an example)

**Review by cblecker — APPROVED** [review]
This looks good to me. Can't find any other meaningful references to google[-_]containers in k/k. Tests are all passing, including cross build.

File: api/openapi-spec/swagger.json (hunk 3 above)
Comment (thockin) [reply api/openapi-spec/swagger.json]: it never really made sense, IMO

**Comment by cblecker** [comment]
/lgtm
/hold

This looks good. Holding to ensure this doesn't merge before @thockin is ready. Feel free to cancel the hold whenever.

**Comment by cblecker** [comment]
/lgtm

@thockin Are you good with this merging whenever tests pass?

**Comment by thockin** [comment]
@cblecker yes

**Comment by cblecker** [comment]
/hold cancel
This should be good then once https://github.com/kubernetes/kubernetes/issues/59223 is fixed.

File: cluster/images/kubemark/Makefile [hunk 4]
```diff
@@ -15,8 +15,8 @@
 # build Kubemark image from currently built binaries containing both 'real' master and Hollow Node.
 # This makefile assumes that the kubemark binary is present in this directory.
 
-REGISTRY?=gcr.io
-PROJECT?=google_containers
+REGISTRY ?= gcr.io
```
Comment (cblecker) [inline cluster/images/kubemark/Makefile]: Did this get lost in a rebase? Should this be staging-k8s.gcr.io?

Comment (thockin) [reply cluster/images/kubemark/Makefile]: ACK, pushed!  Good catch.

**Comment by cblecker** [comment]
/lgtm
Over four hundred lines of the same change starts to blur together after awhile hah. Hopefully this works 🤞. Should be a good time to get it merged too, as the SQ is pretty empty.

File: cluster/images/kubemark/Makefile [hunk 5]
```diff
@@ -15,18 +15,17 @@
 # build Kubemark image from currently built binaries containing both 'real' master and Hollow Node.
 # This makefile assumes that the kubemark binary is present in this directory.
 
-REGISTRY?=gcr.io
-PROJECT?=google_containers
+REGISTRY ?= staging-k8s.gcr.io
```
Comment (cblecker) [inline cluster/images/kubemark/Makefile]: Looks like something is overriding this in testing:
```
I0206 21:23:43.775] gcloud docker -- push gcr.io/kubemark
I0206 21:23:44.496] The push refers to a repository [gcr.io/kubemark]
W0206 21:24:34.697] unknown: Pushing to root-level images is disabled. See https://cloud.google.com/container-registry/docs/support/deprecation-notices
W0206 21:24:34.760] make: *** [gcloudpush] Error 1
```

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/57824/pull-kubernetes-kubemark-e2e-gce/75252/?log#log

Comment (cblecker) [reply cluster/images/kubemark/Makefile]: Maybe it's this: https://github.com/kubernetes/kubernetes/blob/ebffd76acf84d14d7d72fda747e98429d0932895/cluster/images/kubemark/BUILD#L15-L16

Comment (dekkagaijin) [reply cluster/images/kubemark/Makefile]: Doesn't look like `docker_build` messes with env vars:
https://github.com/bazelbuild/rules_docker/blob/master/docker/docker.bzl
https://github.com/bazelbuild/rules_docker/blob/master/container/push.bzl

Comment (BenTheElder) [reply cluster/images/kubemark/Makefile]: adding here: it's not this, that job is using the makefile build. I've poked through test-infra a bit and I don't see this env being set for this job though...

Comment (thockin) [reply cluster/images/kubemark/Makefile]: I know how to fix the makefile part here, I don't know how to fix the bazel part.  Pushing now.  PTAL at bazel equiv.

Comment (thockin) [reply cluster/images/kubemark/Makefile]: pushed another, trying to get bazel.

Bazel doesn't seem to have the same semantic as make here.  Make allows "if this variable is not set, use a default".  So I just repeated the default value.  Sigh.  Someone who knows this script can maybe take a look later ( @shyamjvs )

File: test/kubemark/start-kubemark.sh [hunk 6]
```diff
@@ -221,7 +221,7 @@ function create-and-upload-hollow-node-image {
 function create-and-upload-hollow-node-image-bazel {
   RETRIES=3
   for attempt in $(seq 1 ${RETRIES}); do
-    if ! bazel run //cluster/images/kubemark:push --define PROJECT="${PROJECT}"; then
+    if ! bazel run //cluster/images/kubemark:push --define REGISTRY="${FULL_REGISTRY}"; then
```
Comment (dekkagaijin) [inline test/kubemark/start-kubemark.sh]: This works with the current implementation of docker_build, since it's just concatenating the strings under the covers, but it would be good to be able to be able to assign `$PROJECT/` as a base repository for `kubemark` in the BUILD target, in case that ever changes in the future.

Comment (thockin) [reply test/kubemark/start-kubemark.sh]: ?  PROJECT is a faux construct, AFAICT.  gcr.io/google_containers is the registry.  If you want something else, specify the whole thing.

File: cluster/images/kubemark/BUILD [hunk 7]
```diff
@@ -12,8 +12,8 @@ docker_build(
 docker_push(
     name = "push",
     image = ":image",
-    registry = "gcr.io",
-    repository = "$(PROJECT)/kubemark",
+    registry = "$(REGISTRY)",
+    repository = "kubemark",
```
Comment (cblecker) [inline cluster/images/kubemark/BUILD]: @ixdy @BenTheElder Does this look okay from the bazel side?

Comment (BenTheElder) [reply cluster/images/kubemark/BUILD]: This looks like valid bazel, seems fine to me.

**Comment by BenTheElder** [comment]
/retest
this looks like a known flake, @mikedanese did your timeout bump go in?

**Comment by cblecker** [comment]
@thockin If the test failure is a flake, you should be good to squash

**Comment by BenTheElder** [comment]
All green ✅

File: cluster/images/kubemark/BUILD (hunk 7 above)
Comment (dekkagaijin) [inline cluster/images/kubemark/BUILD]:
```
registry = "$(REGISTRY)",
repository = "$(REPO_PATH_BASE)kubemark",
```

File: test/kubemark/start-kubemark.sh [hunk 8]
```diff
@@ -193,14 +193,14 @@ function create-and-upload-hollow-node-image {
...
-  
+
   echo "Copying kubemark binary to ${MAKE_DIR}"
   cp "${KUBEMARK_BIN}" "${MAKE_DIR}"
   CURR_DIR=`pwd`
   cd "${MAKE_DIR}"
   RETRIES=3
   for attempt in $(seq 1 ${RETRIES}); do
-    if ! REGISTRY="${CONTAINER_REGISTRY}" PROJECT="${PROJECT}" make "${KUBEMARK_IMAGE_MAKE_TARGET}"; then
+    if ! REGISTRY="${FULL_REGISTRY}" make "${KUBEMARK_IMAGE_MAKE_TARGET}"; then
```
Comment (dekkagaijin) [inline test/kubemark/start-kubemark.sh]: `if ! REGISTRY="${REGISTRY}" REPO_PATH_BASE="${REPO_PATH_BASE}" make "${KUBEMARK_IMAGE_MAKE_TARGET}"; then`

File: test/kubemark/start-kubemark.sh [hunk 9]
```diff
@@ -457,6 +457,15 @@ write-local-kubeconfig
...
 
+# TODO: Simplify. The caller should have to pass a full registry or nothing.
+if [[ -n "${CONTAINER_REGISTRY}" && -n "${PROJECT}" ]]; then
+  FULL_REGISTRY="${CONTAINER_REGISTRY}/${PROJECT}"
+elif [[ -n "${CONTAINER_REGISTRY}" ]]; then
+  FULL_REGISTRY="${CONTAINER_REGISTRY}"
+else
+  FULL_REGISTRY=staging-k8s.gcr.io
+fi
+
```
Comment (dekkagaijin) [inline test/kubemark/start-kubemark.sh]:
```
# TODO: Simplify. The caller should have to pass a full registry or nothing.
if [[ -n "${CONTAINER_REGISTRY}" ]]; then
  REGISTRY="${CONTAINER_REGISTRY}"
else
  REGISTRY=staging-k8s.gcr.io
fi
if [[ -n "${PROJECT}" ]]; then
 REPO_PATH_BASE="${PROJECT}/"
fi
```

File: test/kubemark/start-kubemark.sh (hunk 6 above)
Comment (dekkagaijin) [inline test/kubemark/start-kubemark.sh]: `if ! bazel run //cluster/images/kubemark:push --define REGISTRY="${REGISTRY}" --define REPO_PATH_BASE ="${REPO_PATH_BASE}"; then`

Comment (dekkagaijin) [reply test/kubemark/start-kubemark.sh]: In the Docker sense, 'registry' == `hostname[:port]` and 'repository'  == `path` of the URLesque image name: https://docs.docker.com/registry/spec/api/#overview
I'm worried that the a future implementation of `docker_push` might make use of that assumption.

Suggestions added in the relevant places...

Comment (thockin) [reply test/kubemark/start-kubemark.sh]: I prefer it to be simpler - one variable.

I'll leave refactoring of this to the team that owns it ( @shyamjvs and @wojtek-t can overrule if needed)

**Comment by thockin** [comment]
squashed.

**Comment by shyamjvs** [comment]
Oops.. seems like I failed to stop it on time. IIUC this would create problems with our kubemark setup. See point 3 on https://github.com/kubernetes/kubernetes/issues/59567.

**Comment by shyamjvs** [comment]
Sorry, I think I misunderstood the change. You seem to define a FULL_REGISTRY that composes REGISTRY and PROJECT, and those value are taken from [here](https://github.com/kubernetes/kubernetes/blob/3586986416b174d390376311c909d665416bf5c0/test/kubemark/cloud-provider-config.sh#L18-L19) which luckily also continues to take the same previous values [here](https://github.com/kubernetes/kubernetes/blob/3586986416b174d390376311c909d665416bf5c0/test/kubemark/start-kubemark.sh#L398-L399). 

I'll however send a PR to unify these values into one single source of truth to avoid bugs in future.

**Comment by dekkagaijin** [comment]
@shyamjvs would be good to split on the first '/' of the new single source of truth, when passing values to `bazel run //cluster/images/kubemark:push` in `test/kubemark/start-kubemark.sh`:
https://github.com/kubernetes/kubernetes/blob/master/cluster/images/kubemark/BUILD#L12
https://github.com/bazelbuild/rules_docker#container_push
